quda_interface.c

```diff
@@ -1616,97 +1665,86 @@ int invert_eo_MMd_quda(spinor * const out,
-  //  MATDAG_MAT       NORMOP        Solve (A^dag A) x = b
-  
-  //inv_param.solution_type = QUDA_MAT_SOLUTION;   
-  //inv_param.solution_type =QUDA_MATDAG_MAT_SOLUTION  ;  // no
-  //inv_param.solution_type = QUDA_MATPC_SOLUTION ; // no
-  //inv_param.solution_type = QUDA_MATPC_DAG_SOLUTION ;//no
-  inv_param.solution_type = QUDA_MATPCDAG_MATPC_SOLUTION; // # QUDA: ERROR: Source has zero norm
-  //inv_param.solution_type = QUDA_MATPCDAG_MATPC_SHIFT_SOLUTION; //# QUDA: ERROR: Solution type 5 not supported 
-  //inv_param.solution_type = QUDA_MATDAG_MAT_SOLUTION ;
-  //inv_param.solve_type = QUDA_NORMOP_SOLVE ;
-
-  if(solver_params.type == MG || solver_params.type == BICGSTAB){
-    // for MG and BiCGstab, we solve MdagM in two steps
+  if( solver_flag == MG || solver_flag == BICGSTAB){
+    // for MG and BiCGstab, we solve QpQm in two steps
     // we start with M^{-1}
     inv_param.dagger = QUDA_DAG_NO; 
-    inv_param.solve_type = QUDA_DIRECT_PC_SOLVE;
-    inv_param.solution_type = QUDA_MATPC_SOLUTION;
-    if(solver_params.type == MG){
+    if(solver_flag == MG){
       quda_mg_param.invert_param->gamma_basis = QUDA_DEGRAND_ROSSI_GAMMA_BASIS;
     }
   } else {
-    inv_param.solution_type = QUDA_MATPCDAG_MATPC_SOLUTION;
-    inv_param.solve_type = QUDA_NORMOP_PC_SOLVE;
+    // QUDA applies the MMdag operator, we need QpQm^{-1) in the end
+    // so we want QUDA to use the MdagM operator
+    inv_param.dagger = QUDA_DAG_YES; 
   }
-  //inv_param.solve_type = QUDA_NORMERR_PC_SOLVE ;  // QUDA: ERROR: Unpreconditioned MATDAG_MAT solution_type requires an unpreconditioned solve_type
-  inv_param.matpc_type = QUDA_MATPC_ODD_ODD_ASYMMETRIC;
-  //inv_param.matpc_type = QUDA_MATPC_ODD_ODD;
-  //inv_param.matpc_type = QUDA_MATPC_EVEN_EVEN_ASYMMETRIC;
-  //inv_param.matpc_type = QUDA_MATPC_EVEN_EVEN;
 
-  // reorder spinor
   reorder_spinor_eo_toQuda( (double*)spinorIn, inv_param.cpu_prec, 0, 1);
-
-  // perform the inversion
   invertQuda(spinorOut, spinorIn, &inv_param);
-
-  if(solver_flag == MG || solver_flag == BICGSTAB){
+  
+  // the second solve is only necessary in the derivative where we want the inverse of
+  // \hat{Q}^{+} \hat{Q}^{-}
+  // but we're using solvers that don't operate on the normal
+  // system
+  if( (solver_flag == MG || solver_flag == BICGSTAB) && QpQm ){
     inv_param.preserve_source = QUDA_PRESERVE_SOURCE_YES;
-    // for the MG and BiCGstab, we calculate the inverse of MdagM in two steps, now
-    // we do {Mdag}^{-1}
-    inv_param.dagger = QUDA_DAG_YES;
+
+    reorder_spinor_eo_fromQuda( (double*)spinorOut, inv_param.cpu_prec, 0, 1);
+
+    // when using the MG, the symmetrically preconditioned operator is solved
+    // we map to the asymmetric one by multiplying by the inverse of
+    // the diagonal operator on the given parity (not for clover
+    // yet!)
+    if( solver_flag == MG ){
+      mul_one_pm_imu_inv((spinor*)spinorOut, -1, VOLUME/2);
```

Comment:
the sign here is wrong but since it's not working anyway, let's consider it work in progress...